tools/ktx/command_extract.cpp

```diff
@@ -100,468 +70,257 @@ class CommandExtract : public Command {
-            fmt::print(std::cerr, "\n");
-            fmt::print(std::cerr, "{}", validationMessages);
-        }
-
-        if (validationResult != 0)
-            throw FatalError(RETURN_CODE_INVALID_FILE);
-
-        file.seekg(0);
-        if (!file)
-            fatal(RETURN_CODE_IO_FAILURE, "Could not rewind the input file \"{}\": {}", inputFile, errnoMessage());
-    }
+    std::ifstream file(options.inputFilepath, std::ios::in | std::ios::binary);
+    validateToolInput(file, options.inputFilepath, *this);
 
     KTXTexture2 texture{nullptr};
     StreambufStream<std::streambuf*> ktx2Stream{file.rdbuf(), std::ios::in | std::ios::binary};
     auto ret = ktxTexture2_CreateFromStream(ktx2Stream.stream(), KTX_TEXTURE_CREATE_LOAD_IMAGE_DATA_BIT, texture.pHandle());
     if (ret != KTX_SUCCESS)
-        fatal(RETURN_CODE_INVALID_FILE, "Failed to create KTX2 texture: {}", ktxErrorString(ret));
-
-    // TODO Tools P4: Command line argument validation
-    // if (options.level.value_or(0) > texture->numLevels)
-    //     ■ If the "level" option is greater than the maximum level index, generate an error end
-    //             exit.
-    //     fatal(RETURN_CODE_INVALID_ARGUMENTS, )
-    // if (options.layer && !texture->isArray)
-    //     ■ If the texture type is not one of Array types and the "layer" option is set,
-    //             generate an error and exit.
-    // if (...)
-    //     ■ If the texture type is one of Array types and the "layer" option is not set or
-    //             is out of [0, layerCount) range, generate an error and exit.
-    // if (...)
-    //     ■ If the texture type is not Cubemap or Cubemap Array and the "face"
-    //             option is set, generate an error and exit.
-    // if (texture->isCubemap ...)
-    //     ■ If the texture type is Cubemap or Cubemap Array and the "face" option is
-    //             not set or out of [0-5] range, generate an error and exit.
-    // if (...)
-    //     ■ If the texture type is not 3D or 3D Array and the "depth" option is set,
-    //             generate an error and exit.
-
-    std::optional<Swizzle> swizzle;
-    const auto isBasisLZ = texture->supercompressionScheme == KTX_SS_BASIS_LZ;
-
-    if (options.transcode || isBasisLZ) {
-        if (!ktxTexture2_NeedsTranscoding(texture))
-            fatal(RETURN_CODE_INVALID_ARGUMENTS, "Cannot transcode the input file.");
-
-        const auto sample0 = getChannelType(texture, 0);
-        const auto sample1 = getChannelType(texture, 1);
-
-        ktx_transcode_fmt_e targetFormat = KTX_TTF_RGBA32;
-        TranscodeTarget target;
-
-        if (options.transcode)
-            target = *options.transcode;
+        fatal(rc::INVALID_FILE, "Failed to create KTX2 texture: {}", ktxErrorString(ret));
+
+    // CLI request validation
+    if (options.level && options.level != all && options.level >= static_cast<int32_t>(texture->numLevels))
```

Comment:
I have coded the "all" state as -1 as I needed 2 non 'select' state (non-set and select-all) and this was convenient.
It could be improved by a `optional<variant<all_t, uint32_t>>` but thats usage would be cumbersome so I could just create a tiny dedicated `select` type with relevant the accessors.